[ASan] AddressSanitizerPass constructor should honor the AsanCtorKind argument

usama54321 opened the pull request:
Currently, the ConstructorKind member variable in AddressSanitizerPass gets overriden by the ClConstructorKind whether the option is passed from the command line or not.  This override should only happen if the ClConstructorKind argument is passed from the command line.  Otherwise, the constructor should honor the argument passed to it. This patch makes this fix.

rdar://118423755

aeubanks commented on llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp:
```diff
@@ -1149,7 +1152,7 @@ AddressSanitizerPass::AddressSanitizerPass(
     : Options(Options), UseGlobalGC(UseGlobalGC),
       UseOdrIndicator(UseOdrIndicator), DestructorKind(DestructorKind),
-      ConstructorKind(ClConstructorKind) {}
+      ConstructorKind(ConstructorKind) {}
```
should check if the `cl::opt` was set instead of defaulting to `invalid` (can't remember the exact method, `hasValue()`?)

vitalybuka replied:
(Opt.getNumOccurrences() > 0) ? Opt : ConstructorKind;

vitalybuka replied:
and we don't need to introduce Invalid

usama54321 committed:
[ASan] AddressSanitizerPass constructor should honor the AsanCtorKind
argument.

Currently, the ConstructorKind member variable in AddressSanitizerPass
gets overriden by the ClConstructorKind whether the option is passed
from the command line or not.  This override should only happen if the
ClConstructorKind argument is passed from the command line.  Otherwise,
the constructor should honor the argument passed to it. This patch makes
this fix.

rdar://118423755

vitalybuka commented on llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp:
```diff
@@ -1149,7 +1151,7 @@ AddressSanitizerPass::AddressSanitizerPass(
     : Options(Options), UseGlobalGC(UseGlobalGC),
       UseOdrIndicator(UseOdrIndicator), DestructorKind(DestructorKind),
-      ConstructorKind(ClConstructorKind) {}
+      ConstructorKind(ConstructorKind) {}
```
I guess it should be here as well?
```
ConstructorKind(ClConstructorKind.getNumOccurrences() > 0
                            ? ClConstructorKind
                            : ConstructorKind)
```

vitalybuka replied:
I see, It's forwarded there any way

vitalybuka commented:
To properly test this, we need to move DestructorKind and ConstructorKind (and the rest) into AddressSanitizerOptions
and extend parseASanPassOptions. It will make this patch unnecessary complicated. But follow up patches are welcomed.

usama54321 commented:
Thanks @vitalybuka. I will merge this for now and try to follow this up with an AddressSanitizerOptions patch
